iothub/service/src/Amqp/AmqpConnectionHandler.cs

```diff
@@ -250,6 +252,83 @@ private static AmqpSettings CreateAmqpSettings()
             return amqpSettings;
         }
 
+        private async Task<ClientWebSocketTransport> CreateClientWebSocketTransportAsync(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(CreateClientWebSocketTransportAsync));
+
+            try
+            {
+                var websocketUri = new Uri($"{AmqpsConstants.Scheme}{_credential.HostName}:{AmqpsConstants.WebsocketPort}{AmqpsConstants.UriSuffix}");
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, websocketUri, nameof(CreateClientWebSocketTransportAsync));
+
+                ClientWebSocket websocket = CreateClientWebSocket();
+
+                await websocket.ConnectAsync(websocketUri, cancellationToken).ConfigureAwait(false);
+
+                return new ClientWebSocketTransport(websocket, null, null);
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, cancellationToken, nameof(CreateClientWebSocketTransportAsync));
+            }
+        }
+
+        private ClientWebSocket CreateClientWebSocket()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, nameof(CreateClientWebSocket));
+
+            try
+            {
+                // If the user provided a client web socket instance, just use it as is.
+                if (_options.ClientWebSocket != null)
+                {
+                    return _options.ClientWebSocket;
```

Comment:
How do we handle disposing the client web socket?

Do we dispose it no matter what, or only if we created it? I can see customers wanting to share the instance across devices (gateway scenario).

Thoughts?

Reply:
Let's add a flag to the options class that allows users to opt out of us disposing the websocket for them when they dispose the service client

Reply:
Like HttpClient has for disposing HttpClientHandlers

Reply:
I'd rather it not be a separate but related property. Do we bother making a class to hold the bool and the instance?

Reply:
```ClientWebSocketOptions``` class that holds them? I could get behind that

Reply:
That seems like overkill to me

Reply:
I think having a separate option here is fine as long as the docs call out that the SDK will dispose the web socket for them by default and that there is a flag they can set if they want to opt out

Reply:
When you just look at these two properties, it doesn't seem overkill, but complexity comes a little at a time. I'd urge you to make them cohesive.